SDN-5519: OVN-K: Localnet API support (#1750)
openshift/enhancements: enhancements/network/ovn-kubernetes-localnet-api.md

// +kubebuilder:validation:XValidation:rule="has(self.ipam) && has(self.ipam.mode) && self.ipam.mode != 'Enabled' || has(self.subnets)", message="Subnets is required with ipam.mode is Enabled or unset"
// +kubebuilder:validation:XValidation:rule="!has(self.ipam) || !has(self.ipam.mode) || self.ipam.mode != 'Disabled' || !has(self.subnets)", message="Subnets must be unset when ipam.mode is Disabled"
// +kubebuilder:validation:XValidation:rule="!has(self.excludeSubnets) || has(self.excludeSubnets) && !has(self.subnets)", message="excludeSubnets must be unset when subents is unset"
// +kubebuilder:validation:XValidation:rule="!has(self.subnets) || !has(self.mtu) || !self.subnets.exists_one(i, isCIDR(i) && cidr(i).ip().family() == 6) || self.mtu >= 1280", message="MTU should be greater than or equal to 1280 when IPv6 subent is used"
type LocalnetConfig struct {
    // role describes the network role in the pod, required.
    // Whether the pod interface will act as primary or secondary. 
    // For Localnet topology only `Secondary` is allowed.
    // Secondary network is only assigned to pods that use `k8s.v1.cni.cncf.io/networks` annotation to select given network.
    // +required

> 	// +kubebuilder:validation:Enum=Secondary

Reviewer: If `NetworkRole` already specifies an enum validation tag this won't work. To use an Enum tag this needs to be its own type alias to a string or a string. The only other path to restrict this further in this case would be a CEL expression.

Reviewer: using a CEL expression you would end up with a potentially misleading CRD schema though because `Primary` would show up as an allowed value in the schema but in reality it would be blocked by your XValidation

Author: I think we can avoid it by moving the enum valudation for NetworkRole defintion to the toplogies defintions (where NetworkRole used), as follows:
1. Remove NetworkRole enum validation
2. On Layer2Config definition, add  enum validation to `role` field allowing `Primary` or `Secondary`.
3. On Layer3Config definition, add  enum validation to `role` field allowing `Primary` or `Secondary`.
4. Finally, in the proposed  LocalnetConfig definition, add  enum validation to `role` field allowing `Secondary` only.

@everettraven WDYT?

Author: Added text about required changes to topology role validations https://github.com/openshift/enhancements/compare/6fc5b82ece4e9bdaaedf6d7208f56681dc40bcdc..7caa6a8f94524aa3218f9ea2ef845bcc377e4893#diff-65ff4f11b53425ec17e6fbd228843c58470c309f5031c38c345010775bb378d9R508-R519

### Summary
Extend the CUDN CRD to enable creating user-defined networks over localnet topology.
Since the CUDN CRD is targeted for cluster-admin users, it enables preventing non-admin users performing changes that 
could disrupt the cluster or impact the physical network to which the workloads would connect to.

### Localnet using NetworkAttachmentDefinition
As of today OVN-K enables multi-homing including localnet topology networks using NADs
The following NAD YAML describe localnet topology configuration and options:

```yaml
---
apiVersion: k8s.cni.cncf.io/v1
kind: NetworkAttachmentDefinition
metadata:
  name: tenantblue
  namespace: blue
spec:
    config: > '{
        "cniVersion": "0.3.1",
        "type": "ovn-k8s-cni-overlay"
        "netAttachDefName": "blue/tenantblue",
        "topology": "localnet",
        "name": "tenantblue",#1
        "physicalNetworkName": "mylocalnet1", #2
        "subnets": "10.100.0.0/24", #3
        "excludeSubnets": "10.100.50.0/32", #4
        "vlanID": "200", #5
        "mtu": "1500", #6
        "allowPersistentIPs": true #7
    }'
```
1. `name` 
   The underlying network name.
   - Should match the node OVS bridge-mapping network-name.

>    - In case Kubernetes-nmstate is used, should match the `NodeNetworkConfigurationPolicy` (NNCP) `spec.desiredState.ovn.bridge-mappings` value.

Reviewer 2: should actually match the `localnet` attribute of an item on that slice / array.

Author: Added text to clarify that

2. `physicalNetworkName`
   Points to the node OVS bridge-mapping network-name - the network-name mapped to the node OVS bridge that provides access to that network.
   (Can be defined using Kubernetes-nmstate NNCP - `spec.desiredState.ovn.bridge-mappings`)

>    - Overrides the `name` attribute.

Reviewer 2: defined in .1 , for clarity.

Author: Done

The new field should allow users pointing to the bridge-mapping network-name they defined in the node.
The field should be translated to the CNI physicalNetworkName field.

#### MTU
By default OVN-K sets the pod interface with MTU 1400, unless otherwise specified.

For localnet topology this is not optimal because the host physical network-interface usually has default MTU 1500.
In case no MTU is specified the pod interface connected to the host localnet is configured with MTU 1400 resulting 
in bad performance and bad UX requiring troubleshooting the MTU alignment across the stack.

As long OVN-K does not set default MTU (1500) for localnet topology network, 
the controller should specify MTU to be 1500 explicitly in the NAD, if no MTU was specified in the CUDN spec.

### Workflow Description

The CUDN CRD controller should be changes accordingly to support localnet topology.
It should validate localnet topology configuration and generate corresponding NADs for localnet as other topologies (Layer2 & Layer3)
in the selected namespaces.

As of today the CUDN CRD controller does not allow mutating `spec.network`.

OVN-Kubernetes support NAD spec mutations on day 2.

> For the new the spec to take place, restarting the workloads is necessary.

Reviewer 2: nit: 

```suggestion
For the new network configuration (defined in the spec) to take place, the user must restart the workloads.
```

Author: Done

In order to comply with NAD spec mutation support and provide better UX,
the CUDN CRD controller should allow `spec.network` mutations for localnet topologies.
Enable changing at least the following fields: `MTU`, `VLAN`, `excludeSubnets` and `physicalNetworkName`.

In any order:
- The user configures localnet bridge mapping on the nodes, e.g.: using NNCP.
- Create CUDN, specifying the bridge-mapping network name in the spec.

#### Generating the NAD
##### OVS bridge-mapping’s network-name
Introduce attribute to point ovs-bridge bridge-mapping network-name.
This attribute name should be translated to the CNI “physicalNetworkName” attribute.

Proposal the CUDN spec field name:  “physicalNetworkName”.

Alternatively, the field name could be `name`.
In context of the localnet topology configuration it makes sense. For example: 
```yaml
spec:
  network:
    topology: Localnet
    localnet:

>       name: "tenantblue"

Reviewer 2: I still prefer `physicalNetworkName` but, maybe just `name` is good enough.

Author: The more I think about it I think we better stick with ` physicalNetworkName`, I will remove this text.

```

##### MTU
Should be translated to the CNI “mtu” attribute.
If not specified the controller should specify default MTU (1500) in the NAD spec.config.

##### VLAN
Should be translated to the CNI “vlanID” attribute.
If not specified it should not be present in NAD spec.config.

##### Subnets
The subnets and exclude-subnets should be in CIDR form, similar to Layer2 topology subnets.

> ##### Persistent IPs

Reviewer 2: I miss explicitly indicating which is the default. Which I assume to be false.

Author: Yes default is false. Added text to clarify that.

| Subnets             | List of CIDRs used for the pod network across the cluster.Dual-stack clusters may set 2 subnets (one for each IP family), otherwise only 1 subnet is allowed. The format should match standard CIDR notation (for example, "10.128.0.0/16"). This field must be omitted if `ipam.mode` is `Disabled`. | Yes      |
| ExcludeSubnets      | List of CIDRs removed from the specified CIDRs in `subnets`.The format should match standard CIDR notation (for example, "10.128.0.0/16"). This field must be omitted if `subnetes` is unset or `ipam.mode` is `Disabled`.                                                                            | Yes      |
| IPAM                | Contains IPAM-related configuration for the network. When ipam.lifecycle=Persistent enable workloads have persistent IP addresses. For example: Virtual Machines will have the same IP addresses along their lifecycle (stop, start migration, reboots).                                              | Yes      |

#### Suggested API validations 
- `Role`:
  - Required.

>   - Can be Secondary only when topology=Localnet.

Reviewer 2: ```suggestion
  - When topology=Localnet, the only allowed value is `Secondary`.
```

Author: Thank you.
Fixed.

>     - Having Role explicitly makes the API predictable and consistent with other topologies. In addition it enables extending localnet for Primary networks.

Reviewer 2: there's no use case for primary UDN in localnet. It's pointless. 

... but I understand your point.

Author: I changed the text to omit primary role

- `VLAN`: <br/>
  According to [dot1q (IEEE 802.1Q)](https://ieeexplore.ieee.org/document/10004498), 
  VID (VLAN ID) is 12-bits field, providing 4096 values; 0 - 4095. <br/>
  The VLAN IDs `0`, `1` and `4095` are reserved. <br/> 
  Suggested validations:
  - Minimum: 2, Maximum: 4094.
- `Subnets`:
  - Minimum items 1, Maximum items 2.
  - Items are valid CIDR (e.g.: "10.128.0.0/16")
  - When 2 items are specified they must be of different IP families.
- `ExcludeSubnets`:
  - Minimum items 1, Maximum items - 25.
  - Items are valid CIDR (e.g.: "10.128.0.0/16")
  - Cannot be set when Subnet is unset or `ipam.mode=Disabled`.

#### YAML examples
Assuming we have defined NNCP with the following bridge mapping
```yaml
desiredState:
    ovn:
      bridge-mappings:
      - localnet: tenantblue 
        bridge: br-ex
```
Example 1:
```yaml
---
apiVersion: k8s.ovn.org/v1
kind: ClusterUserDefinedNetwork
metadata:
  name: test-net
spec:
  namespaceSelector:
    matchExpressions:
    - key: kubernetes.io/metadata.name
      operator: In
      values: ["red", "blue"]
  network:
    topology: Localnet
    localnet:
      role: Secondary
      physicalNetworkName: tenantblue

>       subnets: ["192.168.100.0/24", "2001:dbb::/64"]

Reviewer 2: for completeness, let's also exclude the GW IPs here (one for each subnet).

Author: Done

Reviewer 2: shouldn't this be a /32 ?

```
The above spec will generate the following NAD, in namespace `blue`:
```yaml
apiVersion: k8s.cni.cncf.io/v1
kind: NetworkAttachmentDefinition
metadata:
  name: test-net
  namespace: blue
finalizers:
 - k8s.ovn.org/user-defined-network-protection
labels:

>   k8s.ovn.org/user-defined-network: ""

Reviewer 2: shouldn't this label be on the namespace ?...

Author: The CRDs controller label NADs with this label making it easier to aggregate for maintenance by admins (view all UDN NAD, etc..).
The primary UDN label for namespace is the same.

Reviewer 2: OK, fair enough.

spec:
    config: > '{
        "cniVersion": "1.0.0",
        "type": "ovn-k8s-cni-overlay"
        "netAttachDefName": "blue/test-net",
        "role": "secondary",
        "topology": "localnet",
        "name": "cluster.udn.test-net",
        "physicalNetworkName: "tenantblue",
        "subnets": "192.168.0.0/16,2001:dbb::/64",
        "allowPersistentIPs": true,
        "excludeSubents: "10.100.50.0/24",
        "mtu": 9000,
        "vlanID": 200
    }'
```

#### Go types
The CUDN `spec.network.topology` field should be extended to accept `Localnet` string.
And the `spec.network` struct `NetworkSpec` should be to have additional field for the localnet topology  configuration:
```go
const NetworkTopologyLocalnet NetworkTopology = "Localnet"

...

// NetworkSpec defines the desired state of UserDefinedNetworkSpec.
// +union
type NetworkSpec struct {
    // Topology describes network configuration.
    //
    // Allowed values are "Layer3", "Layer2" and "Localnet".
    // Layer3 topology creates a layer 2 segment per node, each with a different subnet. Layer 3 routing is used to interconnect node subnets.
    // Layer2 topology creates one logical switch shared by all nodes.

>     // Localnet topology attach to the overlay local network. Enables egress to the provider's physical network. 

Reviewer 2: I kind of always seen this the other way around; Localnet topology (which is an OVN term) attaches to the physical network .

Author: I changed the text saying localnet attach to physical network

    
    ...
    
    // Localnet is the Localnet topology configuration.
    // +optional
    Localnet *LocalnetConfig `json:"localnet,omitempty"`
}
```

The CUDN spec should have additional validation rule for `spec.network.topology` field:
```go
// ClusterUserDefinedNetworkSpec defines the desired state of ClusterUserDefinedNetwork.
type ClusterUserDefinedNetworkSpec struct {
    ...
    // +required
    Network NetworkSpec `json:"network"`
}
```

As of today the CUDN has validation to ensure `spec.network` is immutable:
```go
// +kubebuilder:validation:XValidation:rule="self == oldSelf", message="Network spec is immutable"
```
This validation should be changed in a way it allows mutating localnet topology configurations, 
at least MTU, VLAN, physicalNetworkName and excludeSubnets. 
The proposed change is to invert the immutable-spec validation on CUDN `spec.network` as follows:
1. Remove mutation restrictions on `spec.network`.
2. Add CEL rule validations to restrict mutations of:
   `spec.network.topology`, `spec.network.layer2`, and `spec.network.layer3`.
3. Add CEL rule validations to restrict mutation of localnet topology settings: 

>    `spec.network.localnet.role`, `spec.network.localnet.subnetes` and `spec.network.topology.ipam`.

Reviewer 2: ```suggestion
   `spec.network.localnet.role`, `spec.network.localnet.subnets` and `spec.network.topology.ipam`.
```

Author: Thanks. Fixed.

The CUDN CRD `spec.network` should feature proposed localnet topology configuration.

The Layer2 and Layer3 configuration types (`Layer2Config` & `Layer3Config`) `role` field is defined with `NetworkRole` type.
The `NetworkRole` type has the following enum validation, allowing `Secondary` and `Primary` values:
```
// +kubebuilder:validation:Enum=Primary;Secondary
```
The proposed localnet config type (`LocalnetConfig`) `role` field would have to accept `Secondary` value only.
In order to avoid misleading CRD scheme, the enum validation should be moved closer to each `NetworkRole` usage:
1. Remove the enum validation exist on `NetworkRole` definition
2. At the `Layer2Config` definition, add enum validation to `role` field allowing: `Primary` or `Secondary`.
3. At the `Layer3Config` definition, add enum validation to `role` field allowing: `Primary` or `Secondary`.
4. At the proposed `LocalnetConfig` definition, add enum validation to `role` field allowing `Secondary` only.

```go
// +kubebuilder:validation:XValidation:rule="has(self.ipam) && has(self.ipam.mode) && self.ipam.mode != 'Enabled' || has(self.subnets)", message="Subnets is required with ipam.mode is Enabled or unset"

> // +kubebuilder:validation:XValidation:rule="!has(self.ipam) || !has(self.ipam.mode) || self.ipam.mode != 'Disabled' || !has(self.subnets)", message="Subnets must be unset when ipam.mode is Disabled"

Author: I followed the examples we have today on other topology types, I recall we had a discussion about those rules @npinaeva do you recall by chance why we have the above separated rules?

> // +kubebuilder:validation:XValidation:rule="!has(self.ipam) || !has(self.ipam.mode) || self.ipam.mode != 'Disabled' || !has(self.subnets)", message="Subnets must be unset when ipam.mode is Disabled"

Author: I followed the examples we have today on other topology types for consistency.
I do recall we had a discussion about those rules, from what I remember we wanted to have explicit message describing exactly what went wrong, following recent changes following https://github.com/openshift/enhancements/pull/1725.

 @npinaeva do you recall by chance why we have the above separated rules?

> // +kubebuilder:validation:XValidation:rule="!has(self.ipam) || !has(self.ipam.mode) || self.ipam.mode != 'Disabled' || !has(self.subnets)", message="Subnets must be unset when ipam.mode is Disabled"

Reviewer 3: In other places we set the error message as
```
foo is required when type is Foo, and forbidden otherwise
```

> // +kubebuilder:validation:XValidation:rule="!has(self.ipam) || !has(self.ipam.mode) || self.ipam.mode != 'Disabled' || !has(self.subnets)", message="Subnets must be unset when ipam.mode is Disabled"

Author: Thanks for the feedback :pray: 
Following https://github.com/openshift/enhancements/pull/1750#issuecomment-2682507792, 
I will address this comment in the O-KEP on ovn-kubernetes repo https://github.com/ovn-kubernetes/ovn-kubernetes/pull/5010

// +kubebuilder:validation:XValidation:rule="!has(self.excludeSubnets) || has(self.excludeSubnets) && has(self.subnets)", message="excludeSubnets must be unset when subents is unset"

> // +kubebuilder:validation:XValidation:rule="!has(self.subnets) || !has(self.mtu) || !self.subnets.exists_one(i, isCIDR(i) && cidr(i).ip().family() == 6) || self.mtu >= 1280", message="MTU should be greater than or equal to 1280 when IPv6 subent is used"

Reviewer 3: Exists one checks that there is exactly one item that matches the condition, you here are looking for at least one, in which case you should use `exists`. It is computationally cheaper

Author: The subents field expects the following options: 
1. Single IPv4 CIDR
2. Single IPv6 CIDR 
3. Single IPv4 CIDR and single IPv6 CIDR
In case an IPv6 CIDR present (option 1 or 2) it should ensure MTU is greater then 1280.
Since there should be single item of IPv6 familty we use the `exists_one` operation.
This is true for the existing topologies we have today (Layer2 & Layer3).
Make sense?

Reviewer 3: Ok great, then you don't need `exists_one` here, since you're already checking there's only one IPv6 address

    // +required
    PhysicalNetworkName string `json:"physicalNetworkName"`
    
    // subnets are used for the pod network across the cluster.
    // When set, OVN-Kubernetes assign IP address of the specified CIDRs to the connected pod,
    // saving manual IP assigning or relaying on external IPAM service (DHCP server).
    // subnets is optional, when omitted OVN-Kubernetes won't assign IP address automatically.
    // Dual-stack clusters may set 2 subnets (one for each IP family), otherwise only 1 subnet is allowed.
    // The format should match standard CIDR notation (for example, "10.128.0.0/16").
    // This field must be omitted if `ipam.mode` is `Disabled`.
    // In a scenario `physicalNetworkName` points to OVS bridge mapping of a network who provide IPAM services (e.g.: DHCP server),
    // `ipam.mode` should set with `Disabled, turning off OVN-Kubernetes IPAM and avoid  conflicts with the existing IPAM services on the subject network.
    // +optional
    Subnets DualStackCIDRs `json:"subnets,omitempty"`
    
    // excludeSubnets list of CIDRs removed from the specified CIDRs in `subnets`.

>     // excludeSubnets is optional  When omitted no IP address is excluded and all IP address specified by `subnets` subject to be assigned.

Reviewer 3: ```suggestion
    // excludeSubnets is optional. When omitted no IP address is excluded and all IP address specified by `subnets` subject to be assigned.
```

Author: Fixed

    // The format should match standard CIDR notation (for example, "10.128.0.0/16").
    // This field must be omitted if `subnets` is unset or `ipam.mode` is `Disabled`.
    // In a scenario `physicalNetworkName` points to OVS bridge mapping of a network who has reserved IP addresses
    // that shouldn't be assigned by OVN-Kubernetes, the specified CIDRs will not be assigned. For example:
    // Given: `subnets: "10.0.0.0/24"`, `excludeSubnets: "10.0.0.200/30", the following addresses will not be assigned to pods: `10.0.0.201`, `10.0.0.202`.
    // +optional
    // +kubebuilder:validation:MinItems=1
    // +kubebuilder:validation:MaxItems=25

>     ExcludeSubnets []CIDR `json:"excludeSubnets,omitempty"`

Reviewer 3: You should mention that a maximum of 25 exceptions are allowed

You should also mention that the excluded subnets must be within the range of the `subnets` values. Adding one that isn't contained would be pointless right?

You should also use CEL to validate that

Author: > You should mention that a maximum of 25 exceptions are allowed

> You should also mention that the excluded subnets must be within the range of the subnets values. Adding  one that isn't contained would be pointless right?

Done. 

> You should also use CEL to validate that

I need to figure out how to validate it using CEL.
Do you know of an example that I can use as reference by chance?

Author: Thanks for the reference. 

Following discussion took place on slack, the suggested CEL validation rule doesnt seem to work:
```
self.excludeSubnets.all(e, self.subnets.exists(s,cidr(s).containsCIDR(cidr(e))))
```
It seems the combination between nested CEL macros (all, exists) doesnt seem to play nice with the CIDR operations (containCIDR, cidr).

Nested macros seem to work on simpler example, see this [CEL playground example](https://playcel.undistro.io/?content=H4sIAAAAAAAAA3WS22rCQBCGX2WaXqgQDy30RnojRWiLbaVSpJCbMRnjwmY37OxqpPTdOxsDYkECYTKH%2F%2FtnNz9JTjqZJuMxrEnntiLwFvyO4Gm%2BgKXGY%2BlsMMVNZqTlMgeKAQ0o48lh7tWeRGMzY6Zqo49Q2wM5KoDMXjlrKjI%2BalNTa%2BtIJosYk1Nt5aD87sS1VWUNzJvaEbOScIGmDFgS9IU%2FGEUnrZshrJ3yBEcbXGuNzjM7QXdNX0ytMDpCkFKMWRUEW%2BvEfB08FOgxlRi%2BZ28LkPTr6uM91iv0ncoyKkPvM5heu8UedcArcCxRGfYt6QzohObd%2FnEut1qTnJzM2K0oYFVr4s4Y18phrGXyUJPrUNAqbAx5HqHWfUqBu09qFHvucwoEj8AwGGQmSZNIfYl8ueBbeJYjiVd26erK1qMI7eSnmRnC3WTy30Wbf5icqvF9H3uEW9mCBBn%2FrN8%2FNsGsyWECAAA%3D)

How about the following middle-ground solution: 
The CUDN CRD controller should verify excludeSubets are in range of the specified subents
In a scenario `excludeSubets` are not in range, the controller will raise appropriate error in the status, and not create the corresponding NAD.
@JoelSpeed WDYT?

Author: I filed an issue about it on k8s https://github.com/kubernetes/kubernetes/issues/130441

Reviewer 3: https://github.com/openshift/api/pull/2213 Demonstrates how this would be achieved

Author: We found there is a bug with the CEL validation IP/CIDR library https://github.com/kubernetes/kubernetes/issues/130441#issuecomment-2685185972

How about we relay on the CRD controller validation until the bug is fixed?
Once we have the fix we can enable the validation.

Reviewer 3: Yeah I suspect you'll have to, sadly :( 

What's the supported matrix of kube versions for this project?

Author: We target this feature to 4.19, I suppose it"ll be the kube version used as baseline for OCP 4.19, from what I know it should be 1.32. 

@trozet @tssurya @jcaamano could you please chime in and let us know what should be the kubernetes target version for fixing the mentioned CEL validation rule IP/CIDR library?

Reviewer 3: Yes, we can get the bug into 4.19, but you also need to support upstream Kube versions too right?

Can you say that you need a certain patch version of 1.32/1.31 to upstream clients as the minimum kube version? Not all projects will have that luxury

Reviewer 4: > Yes, we can get the bug into 4.19, but you also need to support upstream Kube versions too right?
> 
> Can you say that you need a certain patch version of 1.32/1.31 to upstream clients as the minimum kube version? Not all projects will have that luxury

by that you mean for OVN-Kube - the client to use the newest api server vendoring right? yes we can make that happen u/s we usually align with what OCP ships

Reviewer 3: It's not about the client, it's about the fact that these CRDs are going to be installed on upstream Kube versions, and this will depend on what the API server is running.

So if you say the latest OVN-Kube version can run on N-2 for example, you'd need to make sure that these validations work on your minimum supported Kube version

  VID (VLAN ID) is 12-bits field, providing 4096 values; 0 - 4095. <br/>
  The VLAN IDs `0`, `1` and `4095` are reserved. <br/> 
  Suggested validations:
  - Minimum: 2, Maximum: 4094.
- `Subnets`:
  - Minimum items 1, Maximum items 2.
  - Items are valid CIDR (e.g.: "10.128.0.0/16")
  - When 2 items are specified they must be of different IP families.
- `ExcludeSubnets`:
  - Minimum items 1, Maximum items - 25.
  - Items are valid CIDR (e.g.: "10.128.0.0/16")
  - Cannot be set when Subnet is unset or `ipam.mode=Disabled`.

#### YAML examples
Assuming we have defined NNCP with the following bridge mapping
```yaml
desiredState:
    ovn:
      bridge-mappings:
      - localnet: tenantblue 
        bridge: br-ex
```
Example 1:
```yaml
---
apiVersion: k8s.ovn.org/v1
kind: ClusterUserDefinedNetwork
metadata:
  name: test-net
spec:
  namespaceSelector:
    matchExpressions:
    - key: kubernetes.io/metadata.name
      operator: In
      values: ["red", "blue"]
  network:
    topology: Localnet
    localnet:
      role: Secondary
      physicalNetworkName: tenantblue
      subnets: ["192.168.100.0/24", "2001:dbb::/64"]

>       excludeSubnets: ["192.168.100.1/31", "2001:dbb::0/128"]

Reviewer 2: don't you want it to be a /32 ?

```go
// +kubebuilder:validation:XValidation:rule="has(self.ipam) && has(self.ipam.mode) && self.ipam.mode != 'Enabled' || has(self.subnets)", message="Subnets is required with ipam.mode is Enabled or unset"
// +kubebuilder:validation:XValidation:rule="!has(self.ipam) || !has(self.ipam.mode) || self.ipam.mode != 'Disabled' || !has(self.subnets)", message="Subnets must be unset when ipam.mode is Disabled"
// +kubebuilder:validation:XValidation:rule="!has(self.excludeSubnets) || has(self.excludeSubnets) && has(self.subnets)", message="excludeSubnets must be unset when subents is unset"
// +kubebuilder:validation:XValidation:rule="!has(self.subnets) || !has(self.mtu) || !self.subnets.exists_one(i, isCIDR(i) && cidr(i).ip().family() == 6) || self.mtu >= 1280", message="MTU should be greater than or equal to 1280 when IPv6 subent is used"
type LocalnetConfig struct {
    // role describes the network role in the pod, required.
    // Whether the pod interface will act as primary or secondary.
    // For Localnet topology only `Secondary` is allowed.
    // Secondary network is only assigned to pods that use `k8s.v1.cni.cncf.io/networks` annotation to select given network.

>     // +kubebuilder:validation:Enum=Secondary

Reviewer: I can't recall if I've mentioned this before, although I feel like I have, but if the `NetworkRole` type also specifies an enum restriction you will not be able to use this validation tag. There are a lot of comments on this PR, so my apologies if we've already discussed this.

    // +required
    Role NetworkRole `json:"role"`
    
    // physicalNetworkName points to the OVS bridge-mapping's network-name configured in the nodes, required.
    // In case OVS bridge-mapping is defined by Kubernetes-nmstate with `NodeNetworkConfigurationPolicy` (NNCP),
    // this field should point to the value of the NNCP spec.desiredState.ovn.bridge-mappings.
    // Max length is 253, cannot contain `,` or `:` characters.

>     // +kubebuilder:validation:MinLength=1

Reviewer: Include in the godoc that the minimum length is 1 (i.e the empty string `""` is not allowed)

Author: Done

    // +kubebuilder:validation:MaxLength=253

>     // +kubebuilder:validation:Pattern="^[^,:]+$"

Reviewer: Conventions have shifted to prefer the use of CEL expressions (the `XValidation` tag) over patterns so that you can provide a more tailored message to your users as opposed to them getting provided the raw regular expression in the admission failure message.
